### PR TITLE
Refactor prompting system

### PR DESCRIPTION
**Pull request by ahuang11**
Starts implementing https://github.com/holoviz/lumen/issues/735

Far from finished, but wanted to gather some feedback first on the changes before updating the other agents too.

One issue I found was that if user provides custom `template_paths` strings (i.e. the template themselves and not paths), I'm not sure it's possible to extend it, so I limited it to path types.

Here, I provide an instructions override:

```python
from lumen.ai.agents import ChatAgent, DuckDBSource
from lumen.ai.memory import memory

memory["current_source"] = DuckDBSource(
    tables=["ibtracs.parquet", "nino_oni.parquet"],
    uri=":memory:",
)

agent = ChatAgent(prompt_overrides={"instructions": "Hello"})
memory["current_data"] = "ABC"
print(await agent._render_system_prompt([]))
```

To the following template:
```jinja2
{% extends 'Agent/main.jinja2' %}

{% block instructions %}
Act as a helpful assistant for high-level data exploration, focusing on available datasets and, only if data is
available, explaining the purpose of each column. Offer suggestions for getting started if needed, remaining factual and
avoiding speculation.
{% endblock %}

{% block context %}
{% if tables|length > 1 %}
Available tables:
{{ closest_tables }}
{% elif schema %}
{{ table }} with schema: {{ schema }}
{% endif %}
{% if memory.get('current_data') %}
Here's a summary of the dataset the user just asked about:
```
{{ memory['current_data'] }}
```
{% endif %}
{% endblock %}
```

Which outputs:
"""

Hello

Available tables:
['ibtracs.parquet', 'nino_oni.parquet']

Here's a summary of the dataset the user just asked about:
```
ABC
```
"""

Note "Act as a helpful assistant for high-level..." is now replaced with "Hello", while the context is untouched.

### CODE REVIEW

File: lumen/ai/agents.py [hunk 1]
```diff
@@ -201,6 +196,21 @@ async def _select_table(self, tables):
...
+        context["memory"] = self._memory
+        system_prompt = render_template(
+            self.template_paths[prompt_name],
+            prompt_overrides=self.prompt_overrides,
+            **context
+        )
+        return system_prompt
+
+    @abstractmethod
+    async def _render_system_prompt(self, messages: list[Message]) -> str:
```
Comment (philippjfr) [inline lumen/ai/agents.py:209]: By system prompt this refers to the "main" prompt?

Comment (ahuang11) [reply lumen/ai/agents.py:209]: Yeah. I suppose I should rename to render_main_prompt

File: lumen/ai/coordinator.py [hunk 2]
```diff
@@ -82,6 +82,17 @@ class Coordinator(Viewer):
     suggestions = param.List(default=GETTING_STARTED_SUGGESTIONS, doc="""
         Initial list of suggestions of actions the user can take.""")
 
+    prompt_templates = param.Dict(default={
```
Comment (philippjfr) [inline lumen/ai/coordinator.py:85]: Should we factor out a baseclass for Agent and Coordinator with shared methods and parameters?

File: lumen/ai/agents.py [hunk 3]
```diff
@@ -20,6 +19,8 @@
 from pydantic import BaseModel, create_model
 from pydantic.fields import FieldInfo
 
+from lumen.ai.llm import Message
```
Comment (philippjfr) [inline lumen/ai/agents.py:22]: Please use relative imports like everywhere else.

**Review by philippjfr — APPROVED** [review]
Soo much better, thank you!

File: lumen/ai/agents.py [hunk 4]
```diff
@@ -798,34 +756,45 @@ class BaseViewAgent(LumenBaseAgent):
...
         )
-        print(f"{self.name} is being instructed that {view_prompt}.")
-
-        # Query
         output = await self.llm.invoke(
             messages,
-            system=system_prompt + view_prompt,
-            response_model=self._get_model(schema),
+            system=system_prompt,
+            response_model=self._output_type,
```
Comment (philippjfr) [inline lumen/ai/agents.py:797]: Not correct, the output_type is the type of the object that is rendered, not the model.

**Pull request merged**